### PR TITLE
Issue548

### PR DESCRIPTION
**Pull request by helloyi**
Belated pull request.

A potential problem is that the metatable of lightuserdata would changed in "shdict", if it set metatable for lightuserdata that anywhere in ngx_lua's VM. The reason is lightuserdata share the metatable globally.

test:
```shell
    $ PATH=${PATH}:/usr/local/openresty/nginx/sbin prove  -I./t/lib  t/126-shdict- frag.t t/043-shdict.t t/080-hup-shdict.t 
    t/126-shdict-frag.t .. ok     
    t/043-shdict.t ....... ok       
    t/080-hup-shdict.t ... ok     
    All tests successful.
    Files=3, Tests=320, 40 wallclock secs ( 0.14 usr  0.01 sys +  4.49 cusr  1.49  csys =  6.13 CPU)
    Result: PASS
```

I find some test-suits rely on external configuration of nginx.
Could you please provide your nginx configuration in `t/servroot`.

Thanks!

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -162,3 +162,7 @@ tthread
 addr2line
 hup
 theaders
+
```
Comment (agentzh) [inline .gitignore]: Hmm, this change is unrelated and should be excluded.

File: src/ngx_http_lua_shdict.c [hunk 2]
```diff
@@ -347,10 +347,13 @@ ngx_http_lua_inject_shdict_api(ngx_http_lua_main_conf_t *lmcf, lua_State *L)
             lua_pushlstring(L, (char *) ctx->name.data, ctx->name.len);
                 /* shared mt key */
 
+            lua_createtable(L, 1 /* narr */, 0 /* nrec */); /* table of zone[i] */
             lua_pushlightuserdata(L, zone[i]); /* shared mt key ud */
+            lua_rawseti(L, -2, 1); /* {zone[i]} */
```
Comment (agentzh) [inline src/ngx_http_lua_shdict.c]: I think it's better to replace the magic number index `1` with a named enum name. For example,

```C
enum {
    SHDICT_USERDATA_INDEX = 1
};
```

near the beginning of this `.c` file and then write

```C
            lua_rawseti(L, -2, SHDICT_USERDATA_INDEX); /* {zone[i]} */
```

What do you think?

File: src/ngx_http_lua_shdict.c [hunk 3]
```diff
@@ -347,10 +347,13 @@ ngx_http_lua_inject_shdict_api(ngx_http_lua_main_conf_t *lmcf, lua_State *L)
...
             lua_pushvalue(L, -3); /* shared mt key ud mt */
             lua_setmetatable(L, -2); /* shared mt key ud */
             lua_rawset(L, -4); /* shared mt */
+                /* name = {zone[i]} */
```
Comment (agentzh) [inline src/ngx_http_lua_shdict.c]: Strange indentation here.

File: src/ngx_http_lua_shdict.c [hunk 4]
```diff
@@ -401,7 +404,13 @@ ngx_http_lua_shdict_get_helper(lua_State *L, int get_stale)
                           "but only seen %d", n);
     }
 
-    zone = lua_touserdata(L, 1);
+    if (LUA_TTABLE != lua_type(L, 1)) {
```
Comment (agentzh) [inline src/ngx_http_lua_shdict.c]: Style: nginx always place constants to be compared on the right hand side of the comparison operator. I do understand the rationale behind this style but modern C compilers already emit this kind of typos (using `=` for `==`) with `-Wall`, like gcc.

File: src/ngx_http_lua_shdict.c [hunk 5]
```diff
@@ -401,7 +404,13 @@ ngx_http_lua_shdict_get_helper(lua_State *L, int get_stale)
...
+        return luaL_error(L, "bad \"zone\" argument");
+    }
+
+    lua_rawgeti(L, 1, 1);
```
Comment (agentzh) [inline src/ngx_http_lua_shdict.c]: We should avoid the magic number index here as well (see above).

File: src/ngx_http_lua_shdict.c [hunk 6]
```diff
@@ -401,7 +404,13 @@ ngx_http_lua_shdict_get_helper(lua_State *L, int get_stale)
...
+    zone = lua_touserdata(L, -1);
```
Comment (agentzh) [inline src/ngx_http_lua_shdict.c]: Better check if the returned `zone` is NULL.

File: src/ngx_http_lua_shdict.c [hunk 7]
```diff
@@ -619,9 +630,11 @@ ngx_http_lua_shdict_flush_expired(lua_State *L)
         return luaL_error(L, "expecting 1 or 2 argument(s), but saw %d", n);
     }
 
-    luaL_checktype(L, 1, LUA_TLIGHTUSERDATA);
+    luaL_checktype(L, 1, LUA_TTABLE);
 
-    zone = lua_touserdata(L, 1);
+    lua_rawgeti(L, 1, 1);
+    zone = lua_touserdata(L, -1);
+    lua_pop(L, 1);
```
Comment (agentzh) [inline src/ngx_http_lua_shdict.c]: Because this code sequence (the 3 lines above) appears frequently, it makes sense to abstract it away with a static (inline) C function like `ngx_http_lua_shdict_get_zone`. What do you think?

**Comment by agentzh** [comment]
@helloyi Thank you for the patch! Will you address the issues in my comments? Thanks!

File: .gitignore (hunk 1 above)
Comment (helloyi) [inline .gitignore]: Haha, yeah, it's personal. This should not happen.

File: src/ngx_http_lua_shdict.c (hunk 2 above)
Comment (helloyi) [inline src/ngx_http_lua_shdict.c]: This is good style. abstract, when necessary.

File: src/ngx_http_lua_shdict.c (hunk 3 above)
Comment (helloyi) [inline src/ngx_http_lua_shdict.c]: Perhaps the code more direct.

File: src/ngx_http_lua_shdict.c (hunk 4 above)
Comment (helloyi) [inline src/ngx_http_lua_shdict.c]: It's really a stale style.

File: src/ngx_http_lua_shdict.c (hunk 7 above)
Comment (helloyi) [inline src/ngx_http_lua_shdict.c]: It's great that taking into account both of readability and efficiency.

**Comment by helloyi** [comment]
Thank you very much for your advice and patient!

File: src/ngx_http_lua_shdict.c [hunk 8]
```diff
@@ -34,12 +34,17 @@ static int ngx_http_lua_shdict_delete(lua_State *L);
 static int ngx_http_lua_shdict_flush_all(lua_State *L);
 static int ngx_http_lua_shdict_flush_expired(lua_State *L);
 static int ngx_http_lua_shdict_get_keys(lua_State *L);
+static inline ngx_shm_zone_t* ngx_http_lua_shdict_get_zone(lua_State *L,
+                                                           int index);
 
 
 #define NGX_HTTP_LUA_SHDICT_ADD         0x0001
 #define NGX_HTTP_LUA_SHDICT_REPLACE     0x0002
 #define NGX_HTTP_LUA_SHDICT_SAFE_STORE  0x0004
```
Comment (agentzh) [inline src/ngx_http_lua_shdict.c]: Style: two blank lines needed here.

File: src/ngx_http_lua_shdict.c [hunk 9]
```diff
@@ -34,12 +34,17 @@ static int ngx_http_lua_shdict_delete(lua_State *L);
...
 
+enum {
+    SHDICT_USERDATA_INDEX = 1,
+};
```
Comment (agentzh) [inline src/ngx_http_lua_shdict.c]: Ditto.

File: src/ngx_http_lua_shdict.c [hunk 10]
```diff
@@ -376,6 +383,17 @@ ngx_http_lua_shdict_get_stale(lua_State *L)
   return ngx_http_lua_shdict_get_helper(L, 1 /* stale */);
 }
```
Comment (agentzh) [inline src/ngx_http_lua_shdict.c]: Ditto.

File: src/ngx_http_lua_shdict.c [hunk 11]
```diff
@@ -376,6 +383,17 @@ ngx_http_lua_shdict_get_stale(lua_State *L)
...
+ngx_http_lua_shdict_get_zone(lua_State *L, int index)
+{
+    ngx_shm_zone_t *zone;
+
+    lua_rawgeti(L, index, SHDICT_USERDATA_INDEX);
+    zone = lua_touserdata(L, -1);
+    lua_pop(L, 1);
+
+    return zone;
+}
```
Comment (agentzh) [inline src/ngx_http_lua_shdict.c]: Ditto.

File: src/ngx_http_lua_shdict.c [hunk 12]
```diff
@@ -376,6 +383,17 @@ ngx_http_lua_shdict_get_stale(lua_State *L)
   return ngx_http_lua_shdict_get_helper(L, 1 /* stale */);
 }
 
+static inline ngx_shm_zone_t *
```
Comment (agentzh) [inline src/ngx_http_lua_shdict.c]: Style: we should use `ngx_inline` instead of `inline` here.

**Comment by agentzh** [comment]
@helloyi Thank you for the quick fixes! Please address the remaining minor style issues :) Thanks!

File: src/ngx_http_lua_shdict.c (hunk 12 above)
Comment (helloyi) [inline src/ngx_http_lua_shdict.c]: Ok! Thanks! :)

**Comment by agentzh** [comment]
@helloyi Looks good to me now. One last thing, sorry, will you please add a couple of test cases for the issue? Maybe we could add the following 2 test cases to `t/043-shdict.t`:

1. check if the `ngx.shared.DICT` value type is a table.
2. check if the lightuserdata `ngx.null` has methods of shared dicts.

We need to ensure these tests indeed fail with the older version though :)

Thanks a lot!

**Comment by helloyi** [comment]
@agentzh I feel that I leant a lot from there, thanks a lot!

**Comment by agentzh** [comment]
@helloyi Great! Looks perfect now :) Thanks for your patience and hard work! I'll merge this thing in the coming days.

File: src/ngx_http_lua_shdict.c [hunk 13]
```diff
@@ -347,7 +356,9 @@ ngx_http_lua_inject_shdict_api(ngx_http_lua_main_conf_t *lmcf, lua_State *L)
             lua_pushlstring(L, (char *) ctx->name.data, ctx->name.len);
                 /* shared mt key */
 
+            lua_createtable(L, 1 /* narr */, 0 /* nrec */); /* table of zone[i] */
```
Comment (agentzh) [inline src/ngx_http_lua_shdict.c]: Style: line exceeding 80 columns. I'll fix it locally during the merge :)

File: src/ngx_http_lua_shdict.c [hunk 14]
```diff
@@ -401,7 +425,11 @@ ngx_http_lua_shdict_get_helper(lua_State *L, int get_stale)
                           "but only seen %d", n);
     }
 
-    zone = lua_touserdata(L, 1);
+    if (lua_type(L, 1) != LUA_TTABLE ) {
```
Comment (agentzh) [inline src/ngx_http_lua_shdict.c]: Style: extra space before the `)` character. I'll fix it locally during the merge anyway.

File: src/ngx_http_lua_shdict.c [hunk 15]
```diff
@@ -36,11 +36,20 @@ static int ngx_http_lua_shdict_flush_expired(lua_State *L);
 static int ngx_http_lua_shdict_get_keys(lua_State *L);
 
 
+static ngx_inline ngx_shm_zone_t* ngx_http_lua_shdict_get_zone(lua_State *L,
```
Comment (agentzh) [inline src/ngx_http_lua_shdict.c]: Style: Line exceeding 80 columns and also we need to swap the `*` and space character right before the function name.

**Comment by agentzh** [comment]
@helloyi It's worth mentioning that we usually use the [ngx-releng](https://github.com/openresty/nginx-devel-utils/blob/master/ngx-releng) tool to catch a few categories of common coding style mistakes automatically for us. Well, just a good trick to save the time ;)

File: src/ngx_http_lua_shdict.c [hunk 16]
```diff
@@ -36,11 +36,20 @@ static int ngx_http_lua_shdict_flush_expired(lua_State *L);
...
+                                                           int index);
```
Comment (agentzh) [inline src/ngx_http_lua_shdict.c]: Style: the parameter declarations need to align with the first argument in the previous line.

File: t/043-shdict.t [hunk 17]
```diff
@@ -2298,3 +2298,100 @@ foo = nil
 --- no_error_log
 [error]
 
+
+
+=== TEST 86: the lightuserdata ngx.null hasn't methods of shared dicts.
```
Comment (agentzh) [inline t/043-shdict.t]: This test does not exercise the bug and also passes successfully with the current master. This is because you did not define any `lua_shared_dict` in this test case, thus no metatables registered on any light user data.

This is why we need to run the tests again the old code version to ensure that they actually fail :) Tests that always pass are meaningless :)

File: t/043-shdict.t [hunk 18]
```diff
@@ -2298,3 +2298,100 @@ foo = nil
...
+    }
+--- request
+GET /test
+--- response_body_like: 500 Internal Server Error
+--- error_code: 500
+--- error_log
+
+
+
+=== TEST 90: value is a table
```
Comment (agentzh) [inline t/043-shdict.t]: This test case is unrelated to the current PR and should be excluded from this patch.

**Comment by agentzh** [comment]
@helloyi A slightly modified version of your latest patch has just been applied to master. Thanks!

**Comment by helloyi** [comment]
@agentzh Thank you very much for the review and merge.
